src/components/EntityHighlighter.js

```diff
@@ -0,0 +1,213 @@
+        this.setState({
+          selectionStart: this.inputNode.selectionStart,
+          selectionEnd: this.inputNode.selectionEnd
+        });
+      }
+    };
+    document.addEventListener('select', this.selectionChangeHandler, false);
+    document.addEventListener('click', this.selectionChangeHandler, false);
+    document.addEventListener('keydown', this.selectionChangeHandler, false);
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('select', this.selectionChangeHandler);
+    document.removeEventListener('click', this.selectionChangeHandler);
+    document.removeEventListener('keydown', this.selectionChangeHandler);
+  }
+
+  handleTextChange(event) {
+    const { text: oldText, entities: oldEntities, onChange } = this.props;
+    const text = event.target.value;
+    const entities = [];
+
+    // update the entity boudaries
+
+    oldEntities.forEach(oldEntity => {
+      const oldSelection = oldText.substr(oldEntity.start, oldEntity.end - oldEntity.start);
+
+      function findClosestStart(lastMatch) {
+        if (lastMatch == null) {
+          const index = text.indexOf(oldSelection);
+          if (index === -1) {
+            return index;
+          }
+          return findClosestStart(index);
+        }
+        const from = lastMatch + oldSelection.length;
+        const index = text.indexOf(oldSelection, from);
+        if (index === -1) {
+          return lastMatch;
+        }
+        const prevDiff = Math.abs(oldEntity.start - lastMatch);
+        const nextDiff = Math.abs(oldEntity.start - index);
+        if (prevDiff < nextDiff) {
+          return lastMatch;
+        }
+        return findClosestStart(index);
+      }
+      const start = findClosestStart();
+      if (start === -1) {
+        return;
+      }
+
+      entities.push({
+        ...oldEntity,
+        start,
+        end: start + oldSelection.length,
+      });
+    });
+
+    onChange(text, entities);
+  }
+
+  focus() {
+    if (this.inputNode) this.inputNode.focus();
+  }
+
+  findEntities = (index) => {
+    return this.props.entities.filter(e => e.start <= index && e.end > index);
+  };
+
+  renderEntityHighlight = (text, entity, key) => {
+    const start = text.substr(0, entity.start);
+    const value = text.substr(entity.start, entity.end - entity.start);
+    const end = text.substr(entity.end);
+    const color = colors[hashString(entity.label) % colors.length].bg;
+    return (
+      <div key={key} style={{ ...styles.zeroPos, ...styles.highlightText }}>
+        <span>{start}</span>
+        <span style={{ opacity: 0.3, backgroundColor: color }}>{value}</span>
+        <span>{end}</span>
+      </div>
+    );
+  };
+
+  deleteEntity = (entity) => {
+    const entities = this.props.entities;
+    const deleted = this.props.entities.findIndex(e => e.start === entity.start && e.end === entity.end && e.label === entity.label);
+    entities.splice(deleted, 1);
+    this.props.onChange(this.props.text, entities);
+  }
```

Comment:
1. :hand: We can pass the entity index from calling code
2. :hand: `splice` would change props array in place, which is not recommended in React, better use e.g. `filter`
```suggestion
  deleteEntity = (entityIndex) => {
    const entities = this.props.entities.filter((entity, idx) => idx !== entityIndex);
    this.props.onChange(this.props.text, entities);
  }
```